### PR TITLE
feat: add control plane metrics library

### PR DESCRIPTION
**Pull request by Xunzhuo**
**What type of PR is this?**

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

After: https://github.com/envoyproxy/gateway/pull/1981
Part of: https://github.com/envoyproxy/gateway/issues/700

### CODE REVIEW

**Review by arkodg — DISMISSED** [review]
LGTM thanks !

File: internal/metrics/metadata.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+// Metric type supports:
+// * Counter: A Counter is a simple metric that only goes up (increments).
+//
+// * Gauge: A Gauge is a metric that represent
+// a single numerical value that can arbitrarily go up and down.
+//
+// * Histogram: A Histogram samples observations and counts them in configurable buckets.
+// It also provides a sum of all observed values.
+// It's used to visualize the statistical distribution of these observations.
```
Comment (zirain) [inline internal/metrics/metadata.go:43]: combine with L32?

Comment (Xunzhuo) [reply internal/metrics/metadata.go:43]: Is this a blocker? Can we do it with a follow-up?

File: internal/metrics/otel_label.go [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package metrics
+
+import "go.opentelemetry.io/otel/attribute"
+
+// A Label provides a named dimension for a Metric.
+type Label struct {
```
Comment (zirain) [inline internal/metrics/otel_label.go:11]: is there any reason not directly use types from OTel?

**Review by zirain — CHANGES_REQUESTED** [review]

File: internal/metrics/otel_label.go (hunk 2 above)
Comment (Xunzhuo) [reply internal/metrics/otel_label.go:11]: We use a wrapper around metrics to hide complexity of otel lib. Take otel gauge as an example, creating a otel gauge for now is complex, we need to implement observable callback. And also other advantages besides easy-to-use, like we can do some optimization but in high level use, we can hide the details.

As for label, why creating a wrapper around it, I don't think we need to import two packages: we need both metrics lib of our own and otel attributes lib, it is weird to be use when creating metrics along with some labels.

File: internal/metrics/register.go [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+		}
+	}
+
+	return nil
+}
+
+// registerOTELgRPCexporter registers OTEL gRPC metrics exporter (PUSH mode).
+func registerOTELgRPCexporter(otelOpts *[]metric.Option, opts registerOptions) error {
+	for _, sink := range opts.pushOptions.sinks {
+		if sink.protocol == "grpc" {
```
[the remark this answers is not in the document]
Comment (Xunzhuo) [reply internal/metrics/register.go:168]: Get it, thanks.

File: internal/metrics/register.go [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+	}
+
+	return nil
+}
+
+// registerOTELHTTPexporter registers OTEL HTTP metrics exporter (PUSH mode).
+func registerOTELHTTPexporter(otelOpts *[]metric.Option, opts registerOptions) error {
+	for _, sink := range opts.pushOptions.sinks {
+		if sink.protocol == "http" {
+			address := fmt.Sprintf("%s:%d", sink.host, sink.port)
```
[the remark this answers is not in the document]
Comment (Xunzhuo) [reply internal/metrics/register.go]: Done.

File: internal/metrics/register.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (Xunzhuo) [reply internal/metrics/register.go]: Done.

File: internal/metrics/otel_metric_counter.go [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+		f.c.Add(context.Background(), value)
+	}
+}
+
+func (f *Counter) Increment() {
+	f.Add(1)
+}
+
+func (f *Counter) Decrement() {
+	f.Add(1)
```
[the remark this answers is not in the document]
Comment (Xunzhuo) [reply internal/metrics/otel_metric_counter.go]: Done.

File: internal/metrics/metadata.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+// Metadata records a metric's metadata.
+type Metadata struct {
+	Name        string
+	Type        MetricType
+	Description string
+	Bounds      []float64
+}
+
+// metrics stores stores metrics
+type metricstore struct {
```
[the remark this answers is not in the document]
Comment (Xunzhuo) [reply internal/metrics/metadata.go]: Done.

**Comment by zirain** [comment]
can you rebase with main branch?

**Comment by Xunzhuo** [comment]
@zirain done.

File: internal/metrics/register.go [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+	}()
+
+	return nil
+}
+
+func newOptions(svr *config.Server) registerOptions {
+	newOpts := registerOptions{}
+	newOpts.address = net.JoinHostPort(v1alpha1.GatewayMetricsHost, fmt.Sprint(v1alpha1.GatewayMetricsPort))
+
+	if !svr.EnvoyGateway.DisablePrometheus() {
```
Comment (zirain) [inline internal/metrics/register.go:79]:
```suggestion
	if svr.EnvoyGateway.DisablePrometheus() {
```
?

Comment (Xunzhuo) [reply internal/metrics/register.go]: Done, thanks.

**Review by zirain — APPROVED** [review]
LGTM, can you add e2e test for this next?

**Review by arkodg — APPROVED** [review]

**Pull request merged**